llmc/adamw.cuh

```diff
@@ -35,15 +34,36 @@ __global__ void adamw_kernel3(Tp* params_memory, float* master_params_memory, Tg
     m /= beta1_correction;  // m_hat
     v /= beta2_correction;  // v_hat
     // fetch the old value of this parameter as a float, from either source
-    float old_param = (master_params_memory != NULL) ? master_params_memory[idx] : (float)params_memory[idx];
-    // update this parameter
+    float old_param;
+    if (mantissas != NULL) {
+        if constexpr (std::is_same_v<Tp, __nv_bfloat16>) {
+            old_param = assemble_float(params_memory[idx], mantissas[idx]);
+        } else {
+            assert(false && "Master params are only implemented for bf16.");
+        }
+    } else {
+        old_param = (float)params_memory[idx];
+    }
+    // update this parameter in 32-bit precision
     float param = old_param - (learning_rate * (m / (sqrtf(v) + eps) + weight_decay * old_param));
+
     // update our low precision version of the parameters using stochastic rounding
     // this will be used in the next forward pass
     // TODO: simply doing `params_memory[i] = (floatX)param;` breaks everything (why?)
-    unsigned int random = Get2dNoiseUint(threadIdx.x, blockIdx.x, seed);
-    stochastic_rounding(param, &params_memory[idx], random);
-    // write the full, float version of the param into our master copy, if we maintain one
-    // this will be used in the next update
-    if (master_params_memory != NULL) { master_params_memory[idx] = param; }
+
+    // If we keep master parameter "copies", make sure to store the missing bits in the 'mantissas' array,
+    // otherwise we can directly go for stochastic rounding.
+    if (mantissas != NULL) {
+        if constexpr (std::is_same_v<Tp, __nv_bfloat16>) {
+            unsigned int random = Get2dNoiseUint(threadIdx.x, blockIdx.x, seed);
```

Comment:
I think this is the only thing that'd have to get modified see the PR I created: https://github.com/karpathy/llm.c/pull/597

Other than this - lgtm!